lib/mongo/retryable.rb

```diff
@@ -46,18 +36,22 @@ module Retryable
     # @return [ Result ] The result of the operation.
     #
     # @since 2.1.0
-    def read_with_retry(attempt = 0, &block)
+    def read_with_retry
+      attempt = 0
       begin
-        block.call
-      rescue Error::SocketError, Error::SocketTimeoutError
-        retry_operation(&block)
+        attempt += 1
+        yield
+      rescue Error::SocketError, Error::SocketTimeoutError => e
+        raise(e) if attempt > cluster.max_read_retries
+        log_retry(e)
+        retry
       rescue Error::OperationFailure => e
         if cluster.sharded? && e.retryable?
-          if attempt < cluster.max_read_retries
+          if attempt <= cluster.max_read_retries
```

Comment:
Why don't we make these lines:

```ruby
raise(e) if attempt > cluster.max_read_retries
sleep(cluster.read_retry_interval)
retry
```

for consistency with the block above.

Also, do we intend to only log the retry if there is a SocketError? Would it not be useful and consistent to also log the retry if there is an OperationFailure?